test/e2e/autoscale_test.go

```diff
@@ -357,12 +383,90 @@ func assertAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods float64, d
 	}
 }
 
+func assertGracefulScaledown(t *testing.T, ctx *testContext, size int) error {
+	// start x running pods; x == size
+	hostConnMap, err := uniqueHostConnections(t, ctx.names, size)
+	if err != nil {
+		return err
+	}
+
+	// only keep openConnCount connections open for the test
+	openConnCount := size / 2
+	deleteHostConnections(hostConnMap, size-openConnCount)
+
+	defer deleteHostConnections(hostConnMap, openConnCount)
+	grp := errgroup.Group{}
+	grp.Go(func() error {
+		timer := time.NewTicker(2 * time.Second)
+		for _ = range timer.C {
+			readyCount, err := numberOfReadyPods(ctx)
+			if err != nil {
+				return err
+			}
+
+			if int(readyCount) < openConnCount {
+				return fmt.Errorf("Failed keeping the right number of pods. Ready(%d) != Expected(%d)", int(readyCount), openConnCount)
+			}
+
+			if int(readyCount) == openConnCount {
+				pods, err := allPods(ctx)
+				if err != nil {
+					return err
+				}
+
+				for _, p := range pods {
+					if p.Status.Phase != corev1.PodRunning || p.DeletionTimestamp != nil {
+						continue
+					}
+
+					if _, ok := hostConnMap.Load(p.Name); !ok {
+						return fmt.Errorf("Failed by keeping the wrong pod %s", p.Name)
+					}
+				}
+				break
+			}
+		}
+		return nil
+	})
+
+	if err := grp.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func TestGracefulScaledown(t *testing.T) {
+	t.Skip()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	ctx := setup(t, autoscaling.KPA, autoscaling.Concurrency, 1 /* target */, 1, /* targetUtilization */
+		wsHostnameTestImageName, nil /* no validation */)
+	defer test.TearDown(ctx.clients, ctx.names)
+
+	autoscalerConfigMap, err := rawCM(ctx.clients, autoscaler.ConfigName)
+	if err != nil {
+		t.Errorf("Error retrieving autoscaler configmap: %v", err)
+	}
+
+	patchedAutoscalerConfigMap := autoscalerConfigMap.DeepCopy()
+	patchedAutoscalerConfigMap.Data["enable-graceful-scaledown"] = "true"
+	patchCM(ctx.clients, patchedAutoscalerConfigMap)
```

Comment:
Will this eventually need to move to a different package or does omitting `Parallel` suffice to make this "safe"?

Reply:
i believe omitting parallel will do the job. that's the recommendation I got from people in the #test slack channel.

Reply:
To be completely safe it requires to be in different package and run separately in the shell script, but I think _this_ particular feature should not affect other tests (and if it does _then_ we do have a problem with the implementation).